src/engine/sync/midimaster.cpp

```diff
@@ -0,0 +1,131 @@
+#include <QtDebug>
+
+#include "engine/sync/midimaster.h"
+
+#include "engine/sync/enginesync.h"
+#include "controllers/midi/midisourceclock.h"
+#include "controllinpotmeter.h"
+#include "controlobject.h"
+#include "controlpushbutton.h"
+#include "configobject.h"
+#include "util/time.h"
+
+MidiMasterClock::MidiMasterClock(const char* pGroup, SyncableListener* pEngineSync)
+        : m_group(pGroup),
+          m_pEngineSync(pEngineSync),
+          m_mode(SYNC_NONE) {
+    m_pMidiSourceClockBpm.reset(new ControlObject(ConfigKey(pGroup, "bpm")));
+    m_pMidiSourceClockLastBeatTime.reset(
+            new ControlObject(ConfigKey(pGroup, "last_beat_time")));
+    m_pMidiSourceClockBeatDistance.reset(
+            new ControlObject(ConfigKey(pGroup, "beat_distance")));
+
+    m_pMidiSourceClockRunning.reset(
+            new ControlPushButton(ConfigKey(pGroup, "run")));
+
+    m_pSyncMasterEnabled.reset(
+            new ControlPushButton(ConfigKey(pGroup, "sync_master")));
+    m_pSyncMasterEnabled->setButtonMode(ControlPushButton::TOGGLE);
+    m_pSyncMasterEnabled->connectValueChangeRequest(
+            this, SLOT(slotSyncMasterEnabledChangeRequest(double)),
+            Qt::DirectConnection);
+
+    m_pMidiSourceClockSyncAdjust.reset(
+            new ControlLinPotmeter(ConfigKey(pGroup, "sync_adjust"), -.5, .5,
+                                   0.1, 0.01, /*allow oob*/ true));
+}
+
+MidiMasterClock::~MidiMasterClock() {
+};
+
+void MidiMasterClock::notifySyncModeChanged(SyncMode mode) {
+    // Syncable has absolutely no say in the matter. This is what EngineSync
+    // requires. Bypass confirmation by using setAndConfirm.
+    m_mode = mode;
+    m_pSyncMasterEnabled->setAndConfirm(mode == SYNC_MASTER);
+}
+
+void MidiMasterClock::notifyOnlyPlayingSyncable() {
+    // No action necessary.
+}
+
+void MidiMasterClock::requestSyncPhase() {
+    // TODO: maybe tell MidiSourceClock to reset which tick is the beat tick??
+    // but really, it's a read-only clock.
+}
+
+void MidiMasterClock::slotSyncMasterEnabledChangeRequest(double state) {
+    bool currentlyMaster = getSyncMode() == SYNC_MASTER;
+
+    if (state > 0.0) {
+        if (currentlyMaster) {
+            // Already master.
+            return;
+        }
+        m_pEngineSync->requestSyncMode(this, SYNC_MASTER);
+    } else {
+        // TODO: midi follower (clock out?)
+        m_pEngineSync->requestSyncMode(this, SYNC_NONE);
+    }
+}
+
+double MidiMasterClock::getBeatDistance() const {
+    qint64 last_beat = static_cast<qint64>(m_pMidiSourceClockLastBeatTime->get());
+    double raw_percent = MidiSourceClock::beatFraction(last_beat, Time::elapsed(),
+                                                       m_pMidiSourceClockBpm->get());
+    raw_percent += m_pMidiSourceClockSyncAdjust->get();
```

Comment:
I am not sure if we are talking about two different things here. 

I have this scenario in mind: 

If we consume a midi clock together with a lets say drum machine, 
It is required that the beat of the drum machine is passed to the external mixer at the same moment the Track in Mixxx passes its beats. 

In Mixxx we have the following delays: 
* Midi receive queue (has not jet a time stamp) 
* quantisation of the audio callback 
* latency of the Mixxx to soundcard DAC 

All this can set for example by playing a track synced by the external master clock in Mixxx and a rhythm on the drum machine. Now you can easily adjust the latency (an additional offset to the external midi clock by ear, (similar to pitch bend) until the famous double base is gone.

Since the latency will to change during a gig, it is clearly a one time setup, that will never cause a lot of problems when playing different tracks with misaligned beat grids or something.  It has nothing to do with the normal sync feature. 

There is only one thing that may go wrong: 
Lets say, the drum machine has a latency of 2 ms and Mixxx has a latency of 50 ms. (Latency = beat is behind the midi tick) The correct correction value is  -48 ms. 
If you run the test with a 180 bpm a valuer of +285 ms a will solve the double base issue as well, but only for this 180 bpm value.
In this case the smaller number is the correct one. But in case the drum machine has lets say 500 ms latency this rule does not apply. 

A call to now() should have always the offset applied, since now() it the future moment when the currently processed sample is converted at the DAC.